server/src/main/java/org/elasticsearch/common/util/CancellableThreads.java

```diff
@@ -185,4 +183,26 @@ public ExecutionCancelledException(StreamInput in) throws IOException {
             super(in);
         }
     }
+
+    /**
+     * Registers a callback that will be invoked when some running operations are cancelled or {@link #checkForCancel()} is called.
+     */
+    public synchronized void setOnCancel(OnCancel onCancel) {
```

Comment:
if you do that is should be using `SetOnce<OnCancel>` to make sure we never override it. The other option is to call it `addOnCancle` and use a list?

Reply:
+1. I pushed https://github.com/elastic/elasticsearch/pull/37291/commits/7c84c0902d672e5ad8203684164f7bd26bc2be49 to use `SetOnce`.